### PR TITLE
Get pokefirered_modern to build

### PR DESCRIPTION
**Pull request by PikalaxALT**
The ROM boots, but crashes at the start of the intro.
pokefirered_modern-0

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -84,7 +88,34 @@ void EnableVCountIntrAtLine150(void);
...
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tstmia r1!, {r0, r3, r4, r5}\n"
+        "\tcmp\tr1, r2\n"
+        "\tbcc\t.LCU0\n"
+    );
```
Comment (jiangzhengwenjz) [inline src/main.c:115]: Again I'm not sure why inline asm is required here. But this looks be better:
```C
    asm(
        "mov\tr1, #0xC0\n"
        "\tlsl\tr1, r1, #0x12\n"
        "\tmov\tr2, #0xFC\n"
        "\tlsl\tr2, r2, #0x7\n"
        "\tadd\tr2, r1, r2\n"
        "\tmov\tr0, #0\n"
        "\tmov\tr3, r0\n"
        "\tmov\tr4, r0\n"
        "\tmov\tr5, r0\n"
        ".LCU%=:\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tstmia\tr1!, {r0, r3, r4, r5}\n"
        "\tcmp\tr1, r2\n"
        "\tbcc\t.LCU%=\n"
        :
        :
        : "r0", "r1", "r2", "r3", "r4", "r5", "memory"
    );
```

File: Makefile [hunk 2]
```diff
@@ -28,16 +72,14 @@ DATA_ASM_BUILDDIR = $(OBJ_DIR)/$(DATA_ASM_SUBDIR)
...
-
-CPPFLAGS := -I tools/agbcc -I tools/agbcc/include -iquote include -nostdinc -undef -D$(GAME_VERSION) -DREVISION=$(GAME_REVISION) -D$(GAME_LANGUAGE)
+ASFLAGS := -mcpu=arm7tdmi --defsym $(GAME_VERSION)=1 --defsym REVISION=$(GAME_REVISION) --defsym $(GAME_LANGUAGE)=1 --defsym MODERN=$(MODERN)
 
 LDFLAGS = -Map ../../$(MAP)
 
-LIB := -L ../../tools/agbcc/lib -lgcc -lc
+LIB := $(LIBPATH) -lgcc -lc
+#ifneq ($(MODERN),0)
+#LIB += -lsysbase
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply Makefile:81]: I was attempting to link the devkitARM libraries but it fell flat.

File: src/main.c (hunk 1 above)
Comment (PikalaxALT) [reply src/main.c:115]: It's not required but it is fast

File: src/main_menu.c [hunk 3]
```diff
@@ -602,7 +602,7 @@ static void PrintMessageOnWindow4(const u8 *str)
     MainMenu_DrawWindow(&sWindowTemplate[4]);
     AddTextPrinterParameterized3(MAIN_MENU_WINDOW_ERROR, 2, 0, 2, sTextColor1, 2, str);
     PutWindowTilemap(MAIN_MENU_WINDOW_ERROR);
-    CopyWindowToVram(MAIN_MENU_WINDOW_ERROR, 2);
+    CopyWindowToVram(MAIN_MENU_WINDOW_ERROR, COPYWIN_GFX);
     SetGpuReg(REG_OFFSET_WIN0H, 0x13DD);
     SetGpuReg(REG_OFFSET_WIN0V, 0x739D);
```
Comment (jiangzhengwenjz) [inline src/main_menu.c:607]: `WIN_RANGE`

File: src/berry_crush_2.c [hunk 4]
```diff
@@ -16,33 +16,33 @@
...
-static u32 sub_814D094(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D188(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D240(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D31C(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D3A4(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D468(struct BerryCrushGame * berryCrushGame, u8 *params);
-static u32 sub_814D4C8(struct BerryCrushGame * berryCrushGame, u8 *params);
-void sub_814D4D8(struct BerryCrushGame * r4);
-void sub_814D5A4(u8 *r0, u32 r1, u32 r2, u32 r3, u32 r5);
+static u32 BerryCrushCommand_BeginNormalPaletteFade(struct BerryCrushGame * game, u8 *r1);
```
Comment (jiangzhengwenjz) [inline src/berry_crush_2.c:19]: original arg names look better and are more consistent

File: src/berry_crush_2.c [hunk 5]
```diff
@@ -479,715 +488,715 @@ static u32 sub_814C24C(struct BerryCrushGame * r4,  UNUSED u8 *r1)
...
 }
 
-void sub_814C618(struct BerryCrushGame * r5)
+static void BerryCrush_HandlePlayerInput(struct BerryCrushGame * game)
 {
-    if (gMain.newKeys & A_BUTTON)
-        r5->unk5C.unk02_2 = 1;
-    if (gMain.heldKeys & A_BUTTON)
+    if (JOY_NEW(A_BUTTON))
+        game->localState.pushedAButton = 1;
```
Comment (jiangzhengwenjz) [inline src/berry_crush_2.c:667]: `TRUE`

File: src/berry_crush_2.c [hunk 6]
```diff
@@ -479,715 +488,715 @@ static u32 sub_814C24C(struct BerryCrushGame * r4,  UNUSED u8 *r1)
...
+    if (JOY_HELD(A_BUTTON))
     {
-        if (r5->unk68.as_four_players.others[r5->unk8].unk4.as_hwords[5] < r5->unk28)
-            ++r5->unk68.as_four_players.others[r5->unk8].unk4.as_hwords[5];
+        if (game->unk68.as_four_players.others[game->localId].unk4.as_hwords[5] < game->timer)
+            ++game->unk68.as_four_players.others[game->localId].unk4.as_hwords[5];
     }
-    if (r5->unk8 != 0 && r5->unk5C.unk02_2 == 0)
+    if (game->localId != 0 && game->localState.pushedAButton == 0)
```
Comment (jiangzhengwenjz) [inline src/berry_crush_2.c:673]: `FALSE`

File: src/berry_crush_2.c [hunk 7]
```diff
@@ -1223,62 +1232,62 @@ static u32 sub_814D188(struct BerryCrushGame * r5, u8 *r6)
...
+        game->cmdState = 0;
         return 0;
     }
-    ++r4->unkC;
+    ++game->cmdState;
     return 0;
 }
 
-static u32 sub_814D31C(struct BerryCrushGame * r5, UNUSED u8 *r1)
+static u32 BerryCrushCommand_FadeOutToPlayAgain(struct BerryCrushGame * r5, UNUSED u8 *r1)
```
Comment (jiangzhengwenjz) [inline src/berry_crush_2.c:1288]: arg names of this and other berry crush cmds can also be updated

File: src/berry_crush_2.c [hunk 8]
```diff
@@ -1350,131 +1359,83 @@ static u32 sub_814D468(struct BerryCrushGame * r5, UNUSED u8 *r1)
...
-    r0[4] = r3;
-    r0[5] = r5;
-    r0[6] = r6;
-    r0[7] = sp[0];
-    r0[8] = sp[1];
-    r0[9] = r1;
+    params[0] = ((u8 *)&selectedPals)[0];
+    params[1] = ((u8 *)&selectedPals)[1];
+    params[2] = ((u8 *)&selectedPals)[2];
+    params[3] = ((u8 *)&selectedPals)[3];
```
Comment (jiangzhengwenjz) [inline src/berry_crush_2.c:1425]: lesson learned

File: src/berry_crush_3.c [hunk 9]
```diff
@@ -1125,82 +1130,81 @@ void ShowBerryCrushRankings(void)
...
 }
 
-static void sub_814EC80(struct BerryCrushGame * r6)
+static void sub_814EC80(struct BerryCrushGame * game)
 {
-    u8 r5 = 0;
+    u8 i = 0;
     const u32 *r0 = gUnknown_8EB0ADC;
-#ifndef NONMATCHING // r4, r5, r6 register roulette
+#ifndef NONMATCHING // r4, i, game register roulette
```
Comment (jiangzhengwenjz) [inline src/berry_crush_3.c:1207]: misrename

**Pull request merged**